src/transfers-history/client.ts

```diff
@@ -0,0 +1,105 @@
+import { ChainId } from "../constants";
+import { clientConfig } from "./config";
+import {
+  DepositEventsQueryServiceFactory,
+  IDepositEventsQueryService,
+  IDepositEventsQueryServiceFactory,
+} from "./services/FundsDepositedEventsQueryService";
+import { TransfersAggregatorService } from "./services/TransfersAggregatorService";
+import { State, TransferFilters, TransferStatus } from "./model/state";
+import { Logger } from "./adapters/logger";
+
+/**
+ * The configuration object for providing the nodes connection details and specifying which
+ * chain should be used as reference (the one that has the highest block time)
+ */
+export type TransfersHistoryClientParams = {
+  chains: {
+    chainId: ChainId;
+    providerUrl: string;
+  }[];
+  refChainId: ChainId;
+};
+
+export class TransfersHistoryClient {
+  constructor(
+    private config: TransfersHistoryClientParams,
+    private state = new State(),
+    private transfersAggregatorService = new TransfersAggregatorService(state),
+    private depositEventsQueryServiceFactory: IDepositEventsQueryServiceFactory = new DepositEventsQueryServiceFactory()
+  ) {
+    for (const chain of config.chains) {
+      clientConfig.web3Providers[chain.chainId] = chain.providerUrl;
+    }
+  }
+
+  public async getTransfers(filters: TransferFilters, limit = 10, offset = 0) {
+    // clean state if the filters change from one call to another
+    if (this.state.filters && this.state.filters !== filters) {
+      this.state.clean();
+    }
+    this.saveFilters(filters);
+
+    const chainsIds = this.config.chains.map(chain => chain.chainId);
+    Logger.debug("[TransfersHistoryClient::getTransfers]", `start getting ${filters.status} transfers from ${chainsIds} chains`);
+    const depositEventsQueryServices = chainsIds
+      .filter(chainId => chainId !== this.config.refChainId)
+      .reduce<[ChainId, IDepositEventsQueryService][]>((acc, chainId) => {
+        return [
+          ...acc,
+          [chainId, this.depositEventsQueryServiceFactory.getService(this.state, chainId, this.config.refChainId, 100000)],
+        ];
+      }, []);
+    const depositEventsRefQueryService = this.depositEventsQueryServiceFactory.getService(
+      this.state,
+      this.config.refChainId,
+      this.config.refChainId,
+      100000
+    );
+    // query events from the chain used as reference
+    await depositEventsRefQueryService.getEvents();
+    await Promise.all(depositEventsQueryServices.map(([_, service]) => service.getEvents()));
+
+    this.transfersAggregatorService.aggregateTransfers();
+    
+    // continue to query events if limit + offset is not reached and we still have blocks to query
+    while (
+      !this.hasRequiredNumberOfTransfersInState(filters.status, limit, offset) &&
+      this.hasBlocksToQuery(chainsIds)
```

Comment:
while maybe a good optimization, i think this adds a lot of complexity, vs just letting everything download in the background while returning to the user everything thats available